raiden/smart_contracts/ChannelManagerLibrary.sol

```diff
@@ -60,31 +61,85 @@ library ChannelManagerLibrary {
     /// @return (address) the address of the NettingChannelContract.
     function newChannel(
         Data storage self,
+        address caller_address,
         address partner,
         uint settle_timeout)
         returns (address)
     {
         address channel_address;
-        uint i;
+        bool has_channel;
+        uint i_index;
+        uint j_index;
 
-        address[] storage existing_channels = self.node_channels[msg.sender];
-        for (i = 0; i < existing_channels.length; i++) {
-            if (NettingChannelContract(existing_channels[i]).partner(msg.sender) == partner) {
-                throw;
-            }
+        (channel_address, has_channel, i_index, j_index) = getChannelWith(self, caller_address, partner);
+        if (!has_channel) {
+            // do nothing, continue to create a new channel
+        } else if (!contractExists(self, channel_address)) {
+            deleteChannel(self, caller_address, partner, channel_address, i_index, j_index);
+        } else {
+            // throw if an open contract exists that is not settled
+            throw;
         }
 
         channel_address = new NettingChannelContract(
             self.token,
-            msg.sender,
+            caller_address,
             partner,
             settle_timeout
         );
 
-        self.node_channels[msg.sender].push(channel_address);
+        self.node_channels[caller_address].push(channel_address);
         self.node_channels[partner].push(channel_address);
         self.all_channels.push(channel_address);
 
         return channel_address;
     }
+
+    /// @notice deleteChannel(address) to remove a channel after it's been settled
+    /// @dev Remove channel after it's been settled
+    /// @param channel_address (address) address of the channel to be closed
+    /// @param partner (address) address of the partner
+    function deleteChannel(
+        Data storage self,
+        address caller_address,
+        address partner,
+        address channel_address,
+        uint i_index,
```

Comment:
Hmm .. since it's also here as a function argument, then please rename as `caller_index` and `partner_index`.
Also add their description in the docstring of the function.